### PR TITLE
stm32U5 spi driver and dma transfer

### PR DESCRIPTION
**Pull request by FRASTM**
This PR is updating the spi driver for the stm32U5 family
and also the DMA drivers for using on spi transfer.

It requires the https://github.com/zephyrproject-rtos/zephyr/pull/45350

The DMA is the GP DMA peripheral of the stm32u5.
The spi1 pins are enabled from the board DTS and two DMA channels are choosen (by the overlay file) from 0 to 11 ( periph to/from SRAM)
The DMA request is selected from the programmed table. 

tested on the disco kit b_u585i_iot02a and nucleo_u575zi_q  board, when D11 and D12 pins (spi MISO <--> MOSI) are connected on the target board.
```
$ west build -p auto -b b_u585i_iot02a -- -DOVERLAY_CONFIG="overlay-stm32-spi.conf" tests/drivers/spi/spi_loopback/
$ west build -p auto -b b_u585i_iot02a -- -DOVERLAY_CONFIG="overlay-stm32-spi-dma.conf" tests/drivers/spi/spi_loopback/
```

Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

### CODE REVIEW

**Comment by FRASTM** [comment]
rebase
requires the https://github.com/zephyrproject-rtos/zephyr/pull/47645

**Comment by FRASTM** [comment]
rebase on d01934f9e9d307f9ca97fae15d5dbc8fa2aaceda

**Comment by FRASTM** [comment]
checks **Run tests with twister / twister-build (7) (pull_request_target)**   failed:
 INFO    - 178/654 nucleo_u575zi_q           samples/drivers/watchdog/sample.drivers.watchdog   FAILED Build failure (build)  is fixed by https://github.com/zephyrproject-rtos/zephyr/pull/48062/

**Comment by FRASTM** [comment]
rebase on https://github.com/zephyrproject-rtos/zephyr/pull/45350 ( https://github.com/zephyrproject-rtos/zephyr/commit/f3017aa787e75c77bb7deb9f4ae45d4b4e93fbc3) to fix CI twister error

**Comment by FRASTM** [comment]
The DMA is now in for the stm32U5 serie, adding the spi with dma transfer
rebase on f033040812b0ce8c5c09105757eb0b3e2e9e8d17

File: drivers/spi/spi_ll_stm32.c [hunk 1]
```diff
@@ -129,6 +129,8 @@ static int spi_stm32_dma_tx_load(const struct device *dev, const uint8_t *buf,
 	/* give the client dev as arg, as the callback comes from the dma */
 	stream->dma_cfg.user_data = data;
 	/* pass our client origin to the dma: data->dma_tx.dma_channel */
+	stream->dma_cfg.linked_channel = STM32_DMA_HAL_OVERRIDE;
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:132]: `STM32_DMA_HAL_OVERRIDE` was designed specifically to be used with drivers using HAL with DMA APIs which conflicts with Zephyr DMA APIS (ex QSPI).
This is not the case here, so it should not be used.
If something specific to enable SPI/DMA for U5 is required it should be done under U5 specific flag and use other mecanism

File: drivers/spi/spi_ll_stm32.c [hunk 2]
```diff
@@ -185,7 +187,7 @@ static int spi_stm32_dma_rx_load(const struct device *dev, uint8_t *buf,
 	/* direction is given by the DT */
 	stream->dma_cfg.head_block = blk_cfg;
 	stream->dma_cfg.user_data = data;
-
+	stream->dma_cfg.linked_channel = STM32_DMA_HAL_OVERRIDE;
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:190]: Ditto

File: drivers/spi/spi_ll_stm32.c [hunk 3]
```diff
@@ -726,7 +728,7 @@ static int transceive_dma(const struct device *dev,
 	/* Set buffers info */
 	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
 
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32U5X)
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:731]: Can we use `DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)` here?
`st,stm32h7-spi` compatible was made specifically to handle these cases specific to H7/U5 variations.

Note: This needs to verify of this is compatible with MP1, which use same compatible

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_ll_stm32.c (hunk 3 above)
Comment (FRASTM) [reply drivers/spi/spi_ll_stm32.c:731]: ok

File: drivers/spi/spi_ll_stm32.c (hunk 1 above)
Comment (FRASTM) [reply drivers/spi/spi_ll_stm32.c]: removed, not needed

File: drivers/spi/spi_ll_stm32.c (hunk 2 above)
Comment (FRASTM) [reply drivers/spi/spi_ll_stm32.c]: removed, not needed

**Comment by FRASTM** [comment]
changed using DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi) also valid for spi node of the stm32u5 serie

**Review by erwango — CHANGES_REQUESTED** [review]
Please update message for the first commit.
The work done is targetting DMA only. U5 SPI support was already there.
Also since `st_stm32h7_spi` compat also applies to MP1, it would be needed to verify and explain in the same commit message that this work is also valid for MP1.

**Comment by FRASTM** [comment]
> Please update message for the first commit. The work done is targetting DMA only. U5 SPI support was already there. Also since `st_stm32h7_spi` compat also applies to MP1, it would be needed to verify and explain in the same commit message that this work is also valid for MP1.

The reason for setting the spi_h7 compatibility also for the stm32u5 and stm32mp1 is around the `LL_SPI_StartMasterTransfer() `because SPI_CFG1 register  "is write protected when SPI is enabled, except TXDMAEN and
RXDMAEN bits".

**Review by erwango — APPROVED** [review]

**Review by ABOSTM — APPROVED** [review]

**Pull request merged**